### ISSUE TITLE
Segfault in ThermodynamicState._find_thermostat()

### ISSUE THREAD

**Issue by jchodera** [choderalab/openmmtools#192]
I just ran into this weird (stochastic?) segfault during a [travis test](https://travis-ci.org/choderalab/openmmtools/jobs/231952029#L2438):
```
ThermodynamicState._find_thermostat() method. ... *** glibc detected *** /home/travis/miniconda/envs/test/bin/python: corrupted double-linked list: 0x00000000131c0830 ***
```

**Comment by jchodera** [comment]
It's not clear if the [failure in `_find_thermostat()`](https://github.com/choderalab/openmmtools/blob/master/openmmtools/states.py#L1304-L1317) is due to a failure in `cls._find_thermostat_index(system)` or in `return system.getForce(thermostat_id)`.

`_find_thermostat_index` has this loop that may be the culprit:
```python
        thermostat_ids = [i for i, force in enumerate(system.getForces())
                          if 'Thermostat' in force.__class__.__name__]
```

**Comment by andrrizzi** [comment]
Yeah, there's a segfault issue that I'm tracking down in a temporary branch, sorry about that.

The failure doesn't seem related to the `_find_thermostat()` method actually. The segfault happens randomly in the first line of the test while copying a `System` shared for the test suite. The C++ memory for the system seems to be deallocated at that point. I think it's an error caused by the combination of nosetest and swig, but I still have to figure out the details.

**Comment by jchodera** [comment]
You're probably much more familiar with this than I am, but these issues seem to often be associated with misplaced Py_DECREF calls in the swig wrapper. Maybe the deepcopy is a bit problematic?

**Comment by andrrizzi** [comment]
Could be! The weird thing is that the system is used and deepcopied in several other tests as well and it works fine, so the last time I spent on this I was trying to figure out which of the previous test ends up corrupting the memory.

**Comment by jchodera** [comment]
valgrind may come in handy if you haven't already resorted to it.
